engine-dmn/engine/pom.xml

```diff
@@ -185,6 +185,25 @@
         </plugins>
       </build>
     </profile>
+    <profile>
+      <id>java16-and-onwards</id>
+      <activation>
+        <jdk>[16,)</jdk>
+      </activation>
+      <build>
+        <plugins>
+          <plugin>
+            <artifactId>maven-surefire-plugin</artifactId>
+            <configuration>
+              <argLine>
+                <!-- org.camunda.bpm.dmn.engine.el.ExpressionCachingTest -->
+                --add-opens java.scripting/javax.script=ALL-UNNAMED
+              </argLine>
+            </configuration>
+          </plugin>
+        </plugins>
+      </build>
+    </profile>
```

Comment:
Instead of opening the package for this test, I suggest we rewrite the test to avoid that we `spy` on JDK classes. This improves the robustness of our tests (we avoid bugs in our production code due to usage of JDK-internal script classes) and we also get closer to Mockito's recommendation to not mock classes we do not own (see https://github.com/mockito/mockito/wiki/How-to-write-good-tests#dont-mock-a-type-you-dont-own and https://github.com/mockito/mockito/issues/2561#issuecomment-1033802654).

I think we can do this the following way:

* Create wrapper classes for the EL provider and ScriptEngine/Compilable  interfaces that count the interactions with expression creation and script compilation
* Instead of using mockito for interaction verification, we can then check the number of invocations from our wrappers
* Instead of spying on `DmnExpressionImpl`, we can just call `#getCachedCompiledScript` directly and verify that it's the same object after both evaluations of the expressions

What do you think? If you like, we can have a brief chat if this approach is unclear.

On a side note: It would be great if you could document the specific reasons why a test like `ExpressionCachingTest` needs such a change. It took me quite a while to figure out that the spying/mocking behavior in Mockito changes (going all the way into `org.mockito.internal.util.reflection.LenientCopyTool.copyValues(T, T, Class<?>)`) depending on the `--add-opens` flag (without it, the copying of fields from the object we are spying to the spy fails).

Reply:
I modified the test. Now it mocks the ScriptEngineManager and ScriptEngine completely instead of spying on them. Which should be fine since the test focuses only on the caching.
No `--add-opens` needed anymore.